logstash-core/lib/logstash/pipelines_registry.rb

```diff
@@ -76,7 +76,6 @@ def put(pipeline_id, state)
     def remove(pipeline_id)
       @lock.synchronize do
         @states.delete(pipeline_id)
-        @locks.delete(pipeline_id)
```

Comment:
please review the locking mechanism.
the idea is to keep `locks[pipeline_id]` forever for `create`, `reload`, `stop` and `delete` mutually exclusive

Reply:
IMO `@locks.delete(pipeline_id)` need to stay there because  keeping `@locks[pipeline_id]` does not make sense if not also keeping `@states[pipeline_id]`.   Once a pipeline is removed, it's `pipeline_id` should not exist anymore in the registry at all. 
I am not sure I understand in which condition it would be useful to keep it around?

Reply:
Thread A, B and C want to update @state of pipeline_id 1 
|  |A|B|C|
|-|-|-|-|
|t1|lock_1 = get_lock(pid: 1)|||
|t2||lock_1 = get_lock(pid: 1)||
|t3|lock_1.lock|||
|t4|update @state|||
|t5|remove @locks[pid: 1]|||
|t6|lock_1.unlock|||
|t7||lock1.lock||
|t8||update @state|lock_1_new = get_lock(pid: 1)|
|t9|||lock_1_new.lock|
|t10|||update @state|

A removes lock in @locks. B holds the old lock
C gets a new lock for pipeline_id 1 as A removed it. B and C have the right to update @state of pipeline_id 1 

I think the purpose of @locks is to ensure only one thread can one pipelind_id state simultaneously.
If @locks keeps the lock, it can keep the integrity of action A,B,C

Reply:
Ha! 💡  I see what you mean. I think you are right here. Let me go over this a bit more.

Reply:
So yes +1 on this change per your reasoning above. The downside here is a memory leak for removed pipelines that will never be recreated but practically speaking the potential for this becoming a problem is extremely low. 
I guess that the alternative would be to rethink/refactor the locking logic but that does not seem necessary. 

I would probably add a comment about this and maybe link it to this discussion here so that if this code is revisited in the future it is more explicit in the code.